feat: Support for additional template variables

chme opened the pull request:
When using a custom Jinja template to render the changelog, it would be great, if additional context information could be accessed in that template. This feature request should be related to #17.

With the changes in this PR, additional context can be passed as key/value pairs either via CLI or via the toml config file.

To pass variables via CLI, the option `--template-var KEY VALUE` can be given multiple times.

To define variables in a configuration file, the variables can be defined in a "[template_vars]"  table.

``` toml
[template_vars]
KEY = "VALUE"
```

When template vars are given via CLI and via config file, they are merged with the CLI vars taking precedence.

All values are accessible in the template through the "template_vars" (dict) variable.

The parsing of the CLI option into a dict is based on this gist: https://gist.github.com/vadimkantorov/37518ff88808af840884355c845049ea

(Unfortunately argparse does not support it out of the box, or at least I could not find a better way to achieve this)

pawamoy reviewed (changes requested):
Amazing feature once again! Thank you so much :blush: 

This all looks good, but I think I would prefer the following syntax on the CLI: `--template-var key=value`. We can do `key_value.split("=", 1)` so that values can have the `=` character too. What do you think?

Also, I don't think I like "template var" for the name and variable/parameter names in the code. What do you think of "jinja context" instead? This way we can also add a short option with `-j` which is not yet taken :slightly_smiling_face: I would then pass this context as `context` or `user_context` when rendering the template ("jinja" would be a bit redundant there).

Finally, it will need some documentation too! I can help with that if needed.

chme commented:
> This all looks good, but I think I would prefer the following syntax on the CLI: `--template-var key=value`. We can do `key_value.split("=", 1)` so that values can have the `=` character too. What do you think?

I first thought, that having key and value as separate parameters, would make it a bit easier, when calling it from a shell script (if the value has to be escaped).
But searching a bit, using the `=` between key and value seems more common (though its difficult to find good examples).

I am fine with `--template-var key=value`.

> Also, I don't think I like "template var" for the name and variable/parameter names in the code. What do you think of "jinja context" instead? This way we can also add a short option with `-j` which is not yet taken 🙂 I would then pass this context as `context` or `user_context` when rendering the template ("jinja" would be a bit redundant there).

Naming is not something I am especially good at :grin:, so whatever you prefer is fine for me.
That said, "template variable" is the term used in the [jinja documentation](https://jinja.palletsprojects.com/en/3.0.x/templates/#variables). 

> Finally, it will need some documentation too! I can help with that if needed.

I can add some documentation (though help on that would be appreciated).

pawamoy commented:
I've pushed some docs :smile: Feel free to tweak them.

chme commented on src/git_changelog/cli.py:
```diff
@@ -100,6 +101,21 @@ def _comma_separated_list(value: str) -> list[str]:
+        option_string: str | Sequence[Any] | None = None,  # noqa: ARG002
+    ):
+        attribute = getattr(namespace, self.dest)
+        if not isinstance(attribute, dict):
+            setattr(namespace, self.dest, {})
```
Defining a default empty dict is not enough. The problem is, when determining whether an argument was explicitly passed via CLI or not:

``` python
    # Determine which arguments were explicitly set with the CLI
    sentinel = object()
    sentinel_ns = argparse.Namespace(**{key: sentinel for key in opts})
    explicit_opts_dict = {
        key: opts.get(key, None)
        for key, value in vars(parser.parse_args(namespace=sentinel_ns, args=args)).items()
        if value is not sentinel
    }
```

During the `parser.parse_args(...)` call, the `_ParseDictAction` is called with the attribute  being an object "special variables". I don't fully understand how this works. Let me know if you know of a better way to handle this.

pawamoy replied:
Right, I see! That's good then :)

pawamoy commented on src/git_changelog/cli.py:
```diff
@@ -487,6 +521,7 @@ def build_and_render(
     bump: str | None = None,
     zerover: bool = True,  # noqa: FBT001,FBT002
     filter_commits: str | None = None,
+    jinja_context: dict[str, str] | None = None,
```
It's true that values passed from the CLI will always be strings, but when using the API directly from Python, values could be anything, so I think it should be `dict[str, Any]`.

pawamoy commented on tests/test_cli.py:
```diff
@@ -212,3 +214,47 @@ def test_show_debug_info(capsys: pytest.CaptureFixture) -> None:
+
+    contents = repo.path.joinpath("CHANGELOG.md").read_text()
+    assert contents == "k1 = v1\nk2 = v2\nk3 = v3\n"
+
+    assert contents == "k1 = v1\nk2 = v2\nk3 = v3\n"
```
```suggestion
```

chme commented:
> I've pushed some docs 😄 Feel free to tweak them.

Awesome, thanks! I only, did a small change in the config file example.

I changed it now to use KEY=VALUE and I hope, I addressed all other comments as well.

pawamoy commented on src/git_changelog/cli.py:
```diff
@@ -458,11 +484,19 @@ def parse_settings(args: list[str] | None = None) -> dict:
     settings.update(explicit_opts_dict)
 
+    # Merge jinja context, CLI values override config file values.
+    if jinja_context and settings.get("jinja_context"):
+        settings["jinja_context"].update(jinja_context)
+    elif jinja_context:
+        settings["jinja_context"] = jinja_context
```
IIUC, `settings["jinja_context"]` is always a dict, so we can safely update it directly.

```suggestion
    settings["jinja_context"].update(jinja_context)
```

pawamoy reviewed (approved):
LGTM, thanks a lot :slightly_smiling_face:
